python_modules/dagster-ext/dagster_ext/__init__.py

```diff
@@ -543,6 +543,80 @@ def upload_messages_chunk(self, payload: IO, index: int) -> None:
 # ########################
 
 
+class IContext(ABC):
+    """Base class for asset context implemented by AssetExecutionContext and ExtContext."""
+
+    @property
+    @abstractmethod
+    def is_asset_step(self) -> bool:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def asset_key(self) -> str:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def asset_keys(self) -> Sequence[str]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def provenance(self) -> Optional[ExtDataProvenance]:
+        """TODO."""
+
+    @property
+    @abstractmethod
+    def provenance_by_asset_key(self) -> Mapping[str, Optional[ExtDataProvenance]]:
```

Comment:
This method does not exist in `OpExecutionContext`, so we would add it to `AssetExecutionContext`

Reply:
Another concern @schrockn  raised is that `provenance` and `code_version` methods reach into the DB for AssetExecutionContext, so making them easy to get is dangerous and encourages users to repeatedly reach out to the db. https://github.com/dagster-io/dagster/pull/16596#discussion_r1329128154